odl/solvers/iterative/iterative.py

```diff
@@ -290,6 +289,106 @@ def conjugate_gradient_normal(op, x, rhs, niter=1, callback=None):
             callback(x)
 
 
+def conjugate_gradient_nonlinear(f, x, rhs, niter=1, nreset=0,
+                                 line_search=None, tol=1e-16, beta_method='FR',
+                                 callback=None):
+    """Conjugate gradient for nonlinear problems.
+
+    The method is described in a
+    `Wikipedia article
+    <https://en.wikipedia.org/wiki/Nonlinear_conjugate_gradient_method>`_.
+
+    Parameters
+    ----------
+    op : `Functional`
+        Operator in the inverse problem. If not linear, it must have
+        an implementation of `Operator.derivative`, which
+        in turn must implement `Operator.adjoint`, i.e.
+        the call ``op.derivative(x).adjoint`` must be valid.
+    x : ``op.domain`` element
+        Vector to which the result is written. Its initial value is
+        used as starting point of the iteration, and its values are
+        updated in each iteration step.
+    rhs : ``op.range`` element
+        Right-hand side of the equation defining the inverse problem
+    niter : int
+        Number of iterations per reset.
+    nreset : int, optional
```

Comment:
If we are interested there are more advanced ways to determine when to reset. The one I know about is by Powell from 1977, but maybe there are more recent ones as well: http://link.springer.com/article/10.1007/BF01593790